jnp.split: use dynamic rather than static slices for speed

jakevdp opened the pull request:
Fixes #12999. See also #9445

This is very similar to the issue & fix in #12219

main branch:
```python
In [1]: import jax.numpy as jnp 
   ...: x = jnp.ones(5000) 
   ...: %time _ = jnp.split(x, jnp.arange(0, 5000, 10))                                                                   
CPU times: user 4.77 s, sys: 118 ms, total: 4.89 s
Wall time: 4.96 s
```
this branch:
```python
In [1]: import jax.numpy as jnp 
   ...: x = jnp.ones(5000) 
   ...: %time _ = jnp.split(x, jnp.arange(0, 5000, 10))                                                                   
CPU times: user 342 ms, sys: 8.52 ms, total: 351 ms
Wall time: 360 ms
```

jakevdp committed:
jnp.split: use dynamic rather than static slices for speed

froystig reviewed (commented):
Do we want benchmarks here, like those you added in #12219?

Does this change affect performance for a _small_ number of splits?

jakevdp commented:
Good question - for repeated operations after the initial run, we get something like this:

main branch:
```python
In [2]: %timeit jax.block_until_ready(jnp.split(x, jnp.arange(0, 5000, 10)))                                          
40.5 ms ± 4.22 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

this branch:
```python
In [2]: %timeit jax.block_until_ready(jnp.split(x, jnp.arange(0, 5000, 10)))                                          
116 ms ± 4.01 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
So a slowdown of ~3x on repeated calls is the tradeoff for saving ~150x on the initial call.

What do you think?

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -1191,9 +1191,11 @@ def _split(op: str, ary: ArrayLike, indices_or_sections: Union[int, ArrayLike],
-  return [lax.slice(ary, _subval(starts, axis, start), _subval(ends, axis, end))
+  # Use dynamic rather than static slice to prevent slow execution of large
+  # number of splits; see https://github.com/google/jax/issues/12999
+  return [lax.dynamic_slice(ary, _subval(starts, axis, start), _subval(sizes, axis, end - start))
```
Another option here: we could replace this with
```python
return [ary[int(start): int(end)] for start, end in zip(split_indices[:-1], split_indices[1:])]
```
Due to #12219, this would now use dynamic rather than static slices and result in the same performance characteristics.

froystig commented:
From conversation with @hawkinsp: we may want to profile in order to understand why the dynamic slice approach is 3x slower, since it isn't clear that it ought to be (?)

KeAWang commented:
Why is `dynamic_slice` faster than static `slice` in the first place?

KeAWang commented:
Thanks! That explains the performance issues I've been trying to debug with slice vs dynamic_slice vs numpy slice.

jakevdp commented:
I can no longer reproduce these performance issues, despite this still lowering to static slice.
